moz_sql_parser/templates/macros.jinja2

```diff
@@ -0,0 +1,227 @@
+{% macro block( content ) %}
+    {% if content|is_list %}
+        {% for field in content %}
+            {% if not loop.first %}
+                {% if not field|is_dict or not ( 'join' in field or 'cross join' in field )%}, {% endif %}
+            {% endif %}
+            {{ block(field) }}
+        {% endfor %}
+    {% elif content|is_dict %}
+
+        {% if 'select' in content or 'from' in content or 'where' in content or 'groupby' in content or 'orderby' in content or 'limit' in content or 'having' in content %}
+            {% if 'select' in content %}
+                select {{ block(content.select) }}
+            {% endif %}
+            {% if 'from' in content %}
+                {% if content.from|is_dict and 'union' in content.from %}
+                {{ block(content.from) }}
+                {% else %}
+                from {{ block(content.from) }}
+                {% endif %}
+            {% endif %}
+            {% if 'where' in content %}
+                where {{ block(content.where) }}
+            {% endif %}
+            {% if 'groupby' in content %}
+                group by {{ block(content.groupby) }}
+            {% endif %}
+            {% if 'having' in content %}
+                having {{ block(content.having) }}
+            {% endif %}
+            {% if 'orderby' in content %}
+                order by {{ block(content.orderby) }}
+            {% endif %}
+            {% if 'limit' in content %}
+                limit {{ block(content.limit) }}
+            {% endif %}
+        {% else %}
+            {% if 'value' in content %}
+                {{ block(content.value) }}{% if 'name' in content %} as {{ content.name }}{% endif %}{% if 'sort' in content %} {{ content.sort }}{% endif %}
+            {% elif 'literal' in content %}
+                {% if content.literal|is_list %}
+                    {% for field in content.literal %}
+                        '{{ field }}'{% if not loop.last %}, {% endif %}
+                    {% endfor %}
+                {% else %}
+                    '{{ content.literal }}'
+                {% endif %}
+            {% elif 'count' in content %}
+                count({{ block(content.count) }})
+            {% elif 'missing' in content %}
+                {{ block(content.missing) }} is NULL
+            {% elif 'exists' in content %}
+                {{ block(content.exists) }} is not NULL
+            {% elif 'join' in content %}
+                join {{ block(content.join) }} on {{ block(content.on) }}
+            {% elif 'when' in content %}
+                when {{ block(content.when) }}
+                {% if 'then' in content %}
+                    then {{ block(content.then) }}
+                {% endif %}
+            {% elif 'case' in content %}
+                case
+                {% if content.case|is_list %}
```

Comment:
There is a lot of this pattern:  `if candidate|is_list`.  If you ensure all candidates are lists then these checks can be removed:

```python
def listwrap(value)
    if value == None:
        return []
    elif isinstance(value, list):
        return value
    elif isinstance(value, set):
        return list(value)
    else:
        return [unwrap(value)]
```